File: backend/src/dto/dto.py
"""
This folder contains the types that will be exposed in our API.
They are known as DTOs (Data Transfer Objects).
"""

from typing import TypedDict, List
from datetime import date

# JWT token
class TokenPayload(TypedDict):
    """
    Represents the decoded JWT payload structure.
    """
    email: str
    account_type: str

class OwnerDTO(TypedDict):
    email: str
    f_name: str
    l_name: str
    province: str
    city: str
    street: str
    phone_num: str
    image_url: str

class EmergencyContactDto(TypedDict):
    phone_num: str
    owner_email: str
    relationship: str
    email: str
    f_name: str
    l_name: str

class DogDTO(TypedDict):
    name: str
    o_email: str
    birth_date: date
    size: str  # Preserve enum type
    image_url: str
    breeds: List[str]

"""
Represents the payload to create a new Dog and its breeds.
"""
class CreateDogDTO(TypedDict):
    name: str
    o_email: str
    birth_date: date
    size: str
    image_filename: str
    breeds: List[str]  # Allow multiple breeds per dog

"""
Represents the payload to update a Dog and its breeds.
"""
class UpdateDogDTO(TypedDict):
    name: str
    o_email: str
    birth_date: date
    size: str
    image_filename: str
    breeds: List[str]  # Allow multiple breeds per dog

class ServiceProviderDTO(TypedDict):
    email: str
    f_name: str
    l_name: str
    province: str
    city: str
    street: str
    phone_num: str
    image_url: str

# Request body shape to create booking
class BookingCreateDto(TypedDict):
    o_email: str
    sp_email: str
    start_datetime: str      # serialized ISO string, not datetime object
    end_datetime: str
    service_type: str        # e.g. "WALKING" or "SITTING"
    price: float
    dog_names: List[str]

File: frontend/src/dto/dto.ts
// Common data transfer objects for API requests

export interface OwnerDto {
  email: string;
  f_name: string;
  l_name: string;
  province: string;
  city: string;
  street: string;
  phone_num: string;
  image_url: string;
}

export interface EmergencyContactDto {
  phone_num: string;
  owner_email: string;
  relationship: string;
  email: string;
  f_name: string;
  l_name: string;
}

export interface DogDto {
  name: string;
  o_email: string;
  birth_date: string; // Serialized ISO string
  size: string;
  image_url: string;
  breeds: Array<string>;
}

export interface BookingDto {
  id: string;
  o_email: string;
  sp_email: string;
  start_datetime: string; // Serialized ISO string
  end_datetime: string; // Serialized ISO string
  service_type: string; // "WALKING" or "SITTING"
  price: number;
  dog_names: Array<string>;
  province: string;
  city: string;
  street: string;
  note: string;
  f_name?: string;
  l_name?: string;
  phone_num?: string;
  image_url?: string;
}

export interface ServiceProviderDto {
  email: string;
  f_name: string;
  l_name: string;
  province: string;
  city: string;
  street: string;
  phone_num: string;
  image_url: string;
}

export interface ReviewDto {
  id: string;
  o_email: string;
  o_image_url: string;
  sp_email: string;
  service_type: string;
  date: Date;
  star_rating: number;
  description?: string;
}
